### ISSUE TITLE
Migrate cephfs and rbd intree plugins by CSI Migration path.

### ISSUE THREAD

**Issue by humblec** [kubernetes/kubernetes#91246]
There are many ceph plugin ( cephfs and rbd)  users and this migration would enable them to get onto ceph csi driver. 

POC PR # https://github.com/kubernetes/kubernetes/pull/95361

**Comment by msau42** [comment]
/remove-lifecycle rotten
@humblec do you plan to work on this in 1.20?

**Comment by humblec** [comment]
@msau42 yeah, I am planning to continue working on this for 1.20 ..

**Comment by humblec** [comment]
@msau42 JFYI I have filed https://github.com/kubernetes/kubernetes/pull/95361 for addressing this, marked it as WIP till I finish some testing on the migration..

**Comment by Jiawei0227** [comment]
Hey @humblec , is there any updates on the CSI migration status for ceph?

**Comment by Jiawei0227** [comment]
Gentle ping. Any updates on this?

**Comment by humblec** [comment]
> Gentle ping. Any updates on this?

@Jiawei0227 would like to discuss few things on this. Can we sync up in slack please ?

**Comment by humblec** [comment]
[Status update] we had some discussions about this internally. The conclusion is that, supporting and maintaining the migration path may not be a worth effort . That said, eventhough we can get the basic operations work ( mount, unmount...etc),  supporting volumes advanced operations ( snapshot, clone, resize..etc) on migrated volumes would be difficult or iow, need some extra design and effort. Also its been some time now we have our Ceph CSI releases in place. 
 In short, unless we have strong requirement from community we may have to mark ceph drivers for deprecation.

@Jiawei0227 will catch up with you in slack and accordingly will take final call here.

**Comment by srhb** [comment]
> The conclusion is that, supporting and maintaining the migration path may not be a worth effort 

Does this mean that operators of clusters that use the in-tree plugin need to design their own migration path?

**Comment by Jiawei0227** [comment]
> supporting volumes advanced operations ( snapshot, clone, resize..etc) on migrated volumes would be difficult or iow

I dont think migrated volume supports snapshot. But resizing should be handled automatically by the translation logic.

I have no objection to deprecate the ceph rbd/ cephfs in-tree plugins. Here is our plan on deprecating the unsupported plugins, I think the most important thing to do here is communication. As long as we are sure all intree ceph plugin users are aware of the change and are switching to CSI model, this should be good.

- 1.22: We will start to mark the plugins as deprecated and create a feature gate to unregister the plugin. But we will keep the feature gate off in this release
- 1.23: We will turn the feature gate on so that if there are users still using in-tree plugins, they can turn off the plugin for now and start to do manual migration to the CSI driver model.
- 1.25: After two releases, we will completely remove the in-tree code and the feature gates. No ceph in-tree plugins usage is allowed at that time

**Comment by msau42** [comment]
I think we need to get more community data on this. Unlike other plugins that have decided to go with the deprecation route, I believe in-tree ceph still has signficant usage in the community, so the impact to users would be significant.

**Comment by msau42** [comment]
Also note that csi migration is about feature parity with in-tree, but not feature parity with CSI. CSI-only features like snapshots, cloning out of scope.

**Comment by humblec** [comment]
> I think we need to get more community data on this. Unlike other plugins that have decided to go with the deprecation route, I believe in-tree ceph still has signficant usage in the community, so the impact to users would be significant.

Agree that we need some data from the community for final call. I was discussing the same and around with @Jiawei0227. Planning to carry out some more discussions on this with different groups and will conclude.

**Comment by sftim** [comment]
There is also https://blog.k8s.io/ and / or https://k8s.dev/blog/ to publicize the CSI driver migration generally and to highlight specific calls for participation like this.

**Comment by jberkus** [comment]
I've pinged some of the Ceph devs about this.

**Comment by humblec** [comment]
The wider communication has been sent out to Kubernetes MLs last week to poll the interest  from the community   about migration/deprecation  of RBD and CephFS plugins.

**Comment by humblec** [comment]
> The wider communication has been sent out to Kubernetes MLs last week to poll the interest from the community about migration/deprecation of RBD and CephFS plugins, meanwhile figuring out the migration effort for CephFS plugin.

Update: Based on the community feedback this feature has been tracked for 1.23 Kube release as Alpha.
